### PR TITLE
Fix #3677: Bugfix in TestSocket `__init__` and `close()`

### LINKED ISSUE #3677: TestSockets are not closed in many tests
I've been investigating issues failures on PyPy and it appears it fails because we open too many pipes during the tests:

image

Why PyPy and not CPython you may ask? Because PyPy has a different `gc` (see [this](https://doc.pypy.org/en/latest/cpython_differences.html)) that doesn't call `__del__` as early as CPython (as soon as the socket is unreferenced), so sockets pile up.

It's easy to find tons of unclosed sockets in the tests (`grep -R TestSocket` then a little bit of looking around)
https://github.com/secdev/scapy/blob/9420c2229bf5330c2cc580f114f63f920a68db10/test/contrib/isotpscan.uts#L72
https://github.com/secdev/scapy/blob/9420c2229bf5330c2cc580f114f63f920a68db10/test/contrib/isotpscan.uts#L163-L164

So we should probably perform some cleanups on those. Feel free to contribute !

### PR DESCRIPTION
**Pull request by polybassa**
This PR fixes issues for #3677

### CODE REVIEW

**Comment by polybassa** [comment]
@gpotter2 Could you please review and merge?

**Comment by polybassa** [comment]
I would like to get some more test runs if this finally solve #3677

**Review by gpotter2 — APPROVED** [review]

File: test/testsocket.py [hunk 1]
```diff
@@ -51,27 +54,30 @@ def close(self):
...
-            for s in self.paired_sockets:
-                try:
-                    s.paired_sockets.remove(self)
-                except (ValueError, AttributeError, TypeError):
-                    pass
-            super(TestSocket, self).close()
+        for s in self.paired_sockets:
             try:
-                open_test_sockets.remove(self)
+                s.paired_sockets.remove(self)
```
[the remark this answers is not in the document]
Comment (polybassa) [reply test/testsocket.py:62]: The idea here is, that every `TestSocket` has a list of `paired_sockets`
https://github.com/secdev/scapy/pull/3698/files/9fa7dcf7d64bc76dc957abc3c3bc2dfa64636a24#diff-7c11f7349558e65920c7157680784fd934cccf3294aa82fca16dd5f061647817R38

During a pair, a `TestSocket` puts itself into the list of `paired_sockets` of the `OtherTestSocket` and the `OtherTestSocket` is stored in its own list of `paired_sockets`.

On `close`, a `TestSocket` removes itself from the list of `paired_sockets` of all `OtherTestSockets`

[the remark this answers is not in the document]
Comment (gpotter2) [reply test/testsocket.py:62]: Okay thanks for the details !

**Comment by evverx** [comment]
@polybassa I found this PR while I was following the trail of https://github.com/secdev/scapy/issues/3880 and I wonder why the test executor uses `time.time` and `time.ctime`? I wonder if it could switch to monotonic timestamps and pass elapsed time around instead of calculating the end point at the beginning. I think should make it a bit more robust in general because it wouldn't be affected by time jumps.

Also I think it would be great if it was possible to fully skip the automotive part when the tests are run by passing `-K automotive` or something like that. This way distributions could just pass it instead of running the test suite on weird architectures (where that part of `scapy` isn't supposed to work as far as I understand).

**Comment by polybassa** [comment]
Thank your very much for pointing this bug out

Most tests (Except ISOTP and CAN) of the automotive layer are actually implemented to have platform independent tests. For performance reasons, I emulate the entire communication with `TestSocket` objects, which are just `queues`. 

But, if you still don't like the automotive tests, most of them can be turned off with the tags `vcan_socket` or `scanner`.

**Comment by evverx** [comment]
> Most tests (Except ISOTP and CAN) of the automotive layer are actually implemented to have platform independent tests.

Got it. Thanks!

> But, if you still don't like the automotive tests

Sorry. I didn't mean to say that I don't like those tests. It's just that given that that part of scapy is somewhat niche I'm not sure whether they should be run on general-purpose distributions and prevent, say, Debian from moving `scapy` to its testing/stable branches if they fail. Generally I'm all for running as many tests as possible to cover as much code as possible.

> most of them can be turned off with the tags vcan_socket or scanner.

I think `-K vcan_socket -K automotive_comm -K scanner` are already passed by the Debian package

**Pull request merged**